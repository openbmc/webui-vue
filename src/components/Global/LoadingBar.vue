<template>
  <transition name="fade">
    <b-progress v-if="!isLoadingComplete" height="0.4rem">
      <b-progress-bar
        striped
        animated
        :value="loadingIndicatorValue"
        :aria-label="$t('global.ariaLabel.progressBar')"
      />
    </b-progress>
  </transition>
</template>

<script>
export default {
  data() {
    return {
      loadingIndicatorValue: 0,
      isLoadingComplete: false,
      loadingIntervalId: null,
      timeoutId: null
    };
  },
  created() {
    this.$root.$on('loader::start', () => {
      this.startLoadingInterval();
    });
    this.$root.$on('loader::end', () => {
      this.endLoadingInterval();
    });
    this.$root.$on('loader::hide', () => {
      this.hideLoadingBar();
    });
  },
  methods: {
    startLoadingInterval() {
      this.clearLoadingInterval();
      this.clearTimeout();
      this.loadingIndicatorValue = 0;
      this.isLoadingComplete = false;
      this.loadingIntervalId = setInterval(() => {
        this.loadingIndicatorValue += 1;
        if (this.loadingIndicatorValue > 100) this.clearLoadingInterval();
      }, 100);
    },
    endLoadingInterval() {
      this.clearLoadingInterval();
      this.clearTimeout();
      this.loadingIndicatorValue = 100;
      this.timeoutId = setTimeout(() => {
        // Let animation complete before hiding
        // the loading bar
        this.isLoadingComplete = true;
      }, 1000);
    },
    hideLoadingBar() {
      this.clearLoadingInterval();
      this.clearTimeout();
      this.loadingIndicatorValue = 0;
      this.isLoadingComplete = true;
    },
    clearLoadingInterval() {
      if (this.loadingIntervalId) clearInterval(this.loadingIntervalId);
      this.loadingIntervalId = null;
    },
    clearTimeout() {
      if (this.timeoutId) clearTimeout(this.timeoutId);
      this.timeoutId = null;
    }
  }
};
</script>

<style lang="scss" scoped>
.progress {
  position: absolute;
  left: 0;
  right: 0;
  bottom: -0.4rem;
  opacity: 1;
  transition: opacity $duration--moderate-01 $standard-easing--productive;

  &.fade-enter,
  &.fade-leave-to {
    opacity: 0;
  }
}
</style>
