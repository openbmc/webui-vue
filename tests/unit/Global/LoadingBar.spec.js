import { mount, createLocalVue } from '@vue/test-utils';
import LoadingBar from '@/components/Global/LoadingBar';
import BootstrapVue from 'bootstrap-vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

describe('LoadingBar.vue', () => {
  const wrapper = mount(LoadingBar, {
    localVue,
    data() {
      return {
        loadingIndicatorValue: 0,
        isLoadingComplete: false,
      };
    },
    mocks: {
      $t: (key) => key,
    },
  });
  it('should exist', () => {
    expect(wrapper.exists()).toBe(true);
  });
  it('should show loading bar element', async () => {
    await wrapper.setData({
      isLoadingComplete: false,
      loadingIndicatorValue: 100,
    });
    expect(wrapper.vm.isLoadingComplete).toBe(false);
    expect(wrapper.find('.progress').exists()).toBe(true);
  });
  it('should hide loading bar element', async () => {
    await wrapper.setData({
      isLoadingComplete: true,
      loadingIndicatorValue: 0,
    });
    expect(wrapper.vm.isLoadingComplete).toBe(true);
    expect(wrapper.find('.progress').exists()).toBe(false);
  });
  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
