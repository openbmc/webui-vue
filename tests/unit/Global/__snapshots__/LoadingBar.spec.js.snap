// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingBar.vue should render correctly 1`] = `
<transition-stub
  name="fade"
>
  <!---->
</transition-stub>
`;
